### ISSUE TITLE
[bug] Apply AdamW weight decay to the pre-update parameters

### ISSUE THREAD

**Issue by tandrewyang** [WangXiangyv/NLPDL-2026Fall#6]
# Body

## Summary

The AdamW pseudocode applies the Adam update first and then applies weight
decay to the already-updated parameter. This is not the decoupled AdamW update
described in the cited paper or implemented by `torch.optim.AdamW`.

## Current behavior

[`README.md`, lines 1496-1504](https://github.com/WangXiangyv/NLPDL-2026Fall/blob/06b82e158bce2c8c374940d0384ecba642b4a9c9/hw1_bpe_and_lm/README.md#L1496-L1504)
specifies the following sequential operations:

```text
theta <- theta - alpha_t * m / (sqrt(v) + epsilon)
theta <- theta - alpha * lambda * theta
```

Because the second line reads the value produced by the first line, the Adam
step is also multiplied by `(1 - alpha * lambda)`. This introduces an extra
cross term and is not merely a floating-point ordering difference.

For example, with `theta=1`, an Adam update of `0.1`, and
`alpha * lambda=0.1`:

```text
standard AdamW:    1 * (1 - 0.1) - 0.1 = 0.80
current pseudocode: (1 - 0.1) * (1 - 0.1) = 0.81
```

The difference, `0.01`, is exactly the unintended cross term.

## Why the public test does not expose it

[`tests/test_optimizer.py`, lines 29-49](https://github.com/WangXiangyv/NLPDL-2026Fall/blob/06b82e158bce2c8c374940d0384ecba642b4a9c9/hw1_bpe_and_lm/tests/test_optimizer.py#L29-L49)
uses `lr=1e-3`, `weight_decay=0.01`, and `atol=1e-4`. Under those values, the
two update rules differ by only about `3e-6` after the supplied 1,000-step
test, so both pass. The test comment also describes the alternatives as
equivalent in principle, but they are algebraically different.

## Impact

A student who implements the published pseudocode exactly can pass the public
test while training with a nonstandard optimizer. The discrepancy becomes
material for larger learning rates or weight decay values and can change the
training trajectory.

## Suggested change

Apply weight decay before the Adam update:

```text
theta <- theta - alpha * lambda * theta
theta <- theta - alpha_t * m / (sqrt(v) + epsilon)
```

Equivalently, combine both terms in one update using the same pre-update
`theta`. A one-step scalar regression test with a larger weight decay would
distinguish the two formulas.
